### PR TITLE
Local native staking impl

### PR DESCRIPTION
**Pull request by maurolacy**
WIP for #10.

TODO:

- ~~Config query~~. (done)
- ~~Max slashing query~~. (done)
- ~~Receive stake~~. (done)
- ~~Release proxy stake~~. (done)
- ~~Proxy by owner query~~. (done)
- ~~Owner by proxy query (uses reverse map)~~. (done)
- Tests.

### CODE REVIEW

File: contracts/native-staking/src/contract.rs [hunk 1]
```diff
@@ -49,8 +60,24 @@ impl NativeStakingContract<'_> {
...
+    fn reply_init_callback(
+        &self,
+        deps: DepsMut,
+        reply: SubMsgResponse,
+    ) -> Result<Response, ContractError> {
+        let init_data = parse_instantiate_response_data(&reply.data.unwrap())?;
+
+        // Associate staking proxy with owner address
+        let proxy_addr = Addr::unchecked(init_data.contract_address);
+        let owner_data: OwnerMsg = from_slice(&init_data.data.unwrap())?; // TODO?: Check for None
```
Comment (hashedone) [inline contracts/native-staking/src/contract.rs:76]: I would avoid unwrapping. Better to raise some missing data or invalid data errors. On the other hand, it is the response to instantiation, so we control the contract ourselves, so neither of those should fire (especially the first one), still - it is called once, and checking (at least the second one) really won't hurt gas-wise, and will report a bug better.

File: contracts/native-staking/src/contract.rs [hunk 2]
```diff
@@ -49,8 +60,24 @@ impl NativeStakingContract<'_> {
...
+        let owner_addr = Addr::unchecked(owner_data.owner);
```
Comment (hashedone) [inline contracts/native-staking/src/contract.rs:77]: This I don't necessarily like. If it would be the address of instantiated address returned from instantiation (here: `proxy_addr`) all would be fine, as this one is guaranteed to be valid by the chain. However, here you have an address set on the response by some external contract, and in case of a bug or accidentally messing up the code id of instantiated contract (maybe some similar contract that just ignored unwanted message fields or malicious contract with code provided by someone) and it can be invalid.

I know that it still can be valid, but not a proper address technically, which we cannot protect from. Still, for better bug detection/reporting, I would prefer to see an invalid address error here than when I try to communicate with it later in the contract.

**Review by hashedone — COMMENTED** [review]
Just some comments after a quick look - I didn't do it carefully, and completely ignored parts that still look to be in progress, but have some suggestions about minorities that might be already useful.

File: contracts/native-staking/src/contract.rs (hunk 2 above)
Comment (maurolacy) [reply contracts/native-staking/src/contract.rs:77]: You mean, add address validation?

The point here is, that this reply handler can only be invoked by this contract itself. I mean, this contract should initiate the entire process, and therefore, it controls the reply's contents, including the payload; as long as this contract is requesting a reply from a known party, that is.

File: contracts/native-staking/src/contract.rs (hunk 1 above)
Comment (ethanfrey) [reply contracts/native-staking/src/contract.rs:76]: Same thing. I prefer `data.ok_or(ContractError::NoInstatiateData)?`

File: contracts/native-staking/src/contract.rs [hunk 3]
```diff
@@ -1,22 +1,32 @@
...
-use crate::types::{Config, ConfigResponse, OwnerByProxyResponse, ProxyByOwnerResponse, StakeMsg};
+use crate::msg::{ConfigResponse, OwnerByProxyResponse, OwnerMsg, ProxyByOwnerResponse, StakeMsg};
+use crate::state::Config;
 
 pub const CONTRACT_NAME: &str = env!("CARGO_PKG_NAME");
 pub const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
 
+pub const REPLY_ID_INSTANTIATE: u64 = 2;
+
+// TODO: Hardcoded for now. Revisit for v1.
```
Comment (ethanfrey) [inline contracts/native-staking/src/contract.rs:23]: This is fair enough.

File: contracts/native-staking/src/contract.rs [hunk 4]
```diff
@@ -49,8 +60,24 @@ impl NativeStakingContract<'_> {
     }
 
     #[msg(query)]
-    fn config(&self, _ctx: QueryCtx) -> Result<ConfigResponse, ContractError> {
-        todo!()
+    fn config(&self, ctx: QueryCtx) -> Result<ConfigResponse, ContractError> {
+        self.config.load(ctx.deps.storage).map_err(Into::into)
```
Comment (ethanfrey) [inline contracts/native-staking/src/contract.rs:64]: Nit: I prefer `Ok(xyz()?)` expression to `xyz().map_err(Into::into)`
They have the same effect, neither is "better". Just my style for what it is worth.

File: contracts/native-staking/src/contract.rs [hunk 5]
```diff
@@ -88,30 +125,55 @@ impl LocalStakingApi for NativeStakingContract<'_> {
...
         ensure_eq!(cfg.vault, ctx.info.sender, ContractError::Unauthorized {});
 
-        // assert funds passed in
+        // Assert funds are passed in
         let _paid = must_pay(&ctx.info, &cfg.denom)?;
 
-        // parse message to find validator to stake on
+        // Parse message to find validator to stake on
         let StakeMsg { validator } = from_slice(&msg)?;
+        // TODO?: Validate validator address
```
Comment (ethanfrey) [inline contracts/native-staking/src/contract.rs:137]: No way to validate (different format, so deps.api can't verify).
Just pass along and it will be validated when staking.

File: contracts/native-staking/src/contract.rs [hunk 6]
```diff
@@ -88,30 +125,55 @@ impl LocalStakingApi for NativeStakingContract<'_> {
...
+                let sub_msg = SubMsg::reply_on_success(msg, REPLY_ID_INSTANTIATE);
+                let owner_data = to_binary(&OwnerMsg { owner })?;
+                Ok(Response::new().add_submessage(sub_msg).set_data(owner_data))
+            }
+            Some(proxy_addr) => {
+                // Send stake message to the proxy contract
+                let msg = WasmMsg::Execute {
+                    contract_addr: proxy_addr.into(),
+                    msg,
+                    funds: ctx.info.funds,
```
Comment (ethanfrey) [inline contracts/native-staking/src/contract.rs:162]: Ah, clever, you never need to look at `_paid` value above, as you just pass along funds

File: contracts/native-staking/src/contract.rs [hunk 7]
```diff
@@ -88,30 +125,55 @@ impl LocalStakingApi for NativeStakingContract<'_> {
...
+        // Look up if there is a proxy to match. Instantiate or call stake on existing
+        match self.proxies.may_load(ctx.deps.storage, &owner_addr)? {
+            None => {
+                // Instantiate proxy contract and send stake message, with reply handling on success
+                let msg = WasmMsg::Instantiate {
+                    admin: Some(ctx.env.contract.address.into()),
+                    code_id: cfg.proxy_code_id,
+                    msg,
+                    funds: ctx.info.funds,
+                    label: format!("LSP for {owner}"), // FIXME: Check / cap label length
```
Comment (ethanfrey) [inline contracts/native-staking/src/contract.rs:151]: This is fine. Labels are like 128 limit. One address is circa 60.

2 addresses in a label may be problematic (if they reference contracts not external accounts)

File: contracts/native-staking/src/contract.rs [hunk 8]
```diff
@@ -88,30 +125,55 @@ impl LocalStakingApi for NativeStakingContract<'_> {
...
+            }
+            Some(proxy_addr) => {
+                // Send stake message to the proxy contract
+                let msg = WasmMsg::Execute {
+                    contract_addr: proxy_addr.into(),
+                    msg,
+                    funds: ctx.info.funds,
+                };
+                let sub_msg = SubMsg::new(msg);
+                Ok(Response::new().add_submessage(sub_msg))
```
Comment (ethanfrey) [inline contracts/native-staking/src/contract.rs:165]: Nit: `Ok(Response::new().add_message(msg))`
does the exact same thing without the line above

File: contracts/native-staking/src/contract.rs [hunk 9]
```diff
@@ -88,30 +125,55 @@ impl LocalStakingApi for NativeStakingContract<'_> {
...
+        let owner_addr = ctx.deps.api.addr_validate(&owner)?;
+
+        // Look up if there is a proxy to match. Instantiate or call stake on existing
+        match self.proxies.may_load(ctx.deps.storage, &owner_addr)? {
+            None => {
+                // Instantiate proxy contract and send stake message, with reply handling on success
+                let msg = WasmMsg::Instantiate {
+                    admin: Some(ctx.env.contract.address.into()),
+                    code_id: cfg.proxy_code_id,
+                    msg,
```
Comment (ethanfrey) [inline contracts/native-staking/src/contract.rs:149]: You don't want to blindly pass msg.
You need to build a new message (with denom, owner, validator). 
See [proxy design](https://github.com/osmosis-labs/mesh-security/blob/main/contracts/native-staking-proxy/src/contract.rs#L30-L38)

2 points:
1. It would be nice to have auto-generated helpers to call this type-safe. I mean `native_staking_proxy::contract::InstantiateMsg` _should_ be the right type. I was proposing to make helper functions to build this next level. (And @hashedone can give a tip on how to call type-safe)
2. If you think `denom` is unneeded, let's remove it from the message. We can probably get it by querying `StakingQuery::BondDenom{}` or such. But if removed, I'd remove it from the instantiate method of this contract as well.

**Review by ethanfrey — COMMENTED** [review]
Nice start, added some comments.

File: contracts/native-staking/src/contract.rs (hunk 5 above)
Comment (JakeHartnell) [reply contracts/native-staking/src/contract.rs:137]: Yeah, went through a [long discussion with this previously on the prototype](https://github.com/CosmWasm/mesh-security-hackathon/issues/42). FWIW, some newer Cosmos chains like Penumbra are also ditching `valoper` addresses (not to mention other non-SDK cosmwasm VMs like Composable), so not sure it's worth validating in the contract as it's done on the SDK or native chain code.

File: contracts/native-staking/src/contract.rs (hunk 9 above)
Comment (maurolacy) [reply contracts/native-staking/src/contract.rs:149]: Good catch. Fixed it.

File: contracts/native-staking/src/contract.rs [hunk 10]
```diff
@@ -88,30 +139,62 @@ impl LocalStakingApi for NativeStakingContract<'_> {
...
+                let wasm_msg = WasmMsg::Instantiate {
+                    admin: Some(ctx.env.contract.address.into()),
+                    code_id: cfg.proxy_code_id,
+                    msg,
+                    funds: ctx.info.funds,
+                    label: format!("LSP for {owner}"),
+                };
+                let sub_msg = SubMsg::reply_on_success(wasm_msg, REPLY_ID_INSTANTIATE);
+                let owner_data = to_binary(&OwnerMsg { owner })?;
+                Ok(Response::new().add_submessage(sub_msg).set_data(owner_data))
```
Comment (ethanfrey) [inline contracts/native-staking/src/contract.rs:175]: Why setting data here? I mean the caller knows `owner` already.

If you want to return the `proxy` address, that might make sense to save a later query. But totally optional as it can be queried.

File: contracts/native-staking/src/contract.rs [hunk 11]
```diff
@@ -88,30 +139,62 @@ impl LocalStakingApi for NativeStakingContract<'_> {
...
-        let _ = validator;
-
-        let _ = owner;
 
-        // look up if there is a proxy to match
-        // instantiate or call stake on existing
-        todo!();
+        let owner_addr = ctx.deps.api.addr_validate(&owner)?;
+
+        // Look up if there is a proxy to match. Instantiate or call stake on existing
```
Comment (ethanfrey) [inline contracts/native-staking/src/contract.rs:154]: Method looks good in general

File: contracts/native-staking/src/contract.rs [hunk 12]
```diff
@@ -123,10 +206,21 @@ impl NativeStakingCallback for NativeStakingContract<'_> {
...
-        // look up proxy address (info.sender) to account owner
-        // send these tokens to vault contract, using release_local_stake method
-        todo!()
+    fn release_proxy_stake(&self, ctx: ExecCtx) -> Result<Response, Self::Error> {
+        let cfg = self.config.load(ctx.deps.storage)?;
+
+        // Assert funds are passed in
+        let _paid = must_pay(&ctx.info, &cfg.denom)?;
+
+        // Look up account owner by proxy address (info.sender)
```
Comment (ethanfrey) [inline contracts/native-staking/src/contract.rs:215]: Add a note that this asserts the caller is a valid proxy

**Review by ethanfrey — APPROVED** [review]
First of all, nice work.

It is unfortunate, that this is based on vault. It would be good to get smaller independent PRs merged often to main.

In particularly now, we need:
1. Tests on this and #30 are both blocked on a new Sylvia release (likely tomorrow)
2. I am unclear when vault will be merged
3. native-staking-proxy needs a new PR

I would actually be happy to have both merged when they compile and pass CI, but missing tests. Then do tests in a new PR (I say this here as there is sylvia changes needed for tests and getting them in now may avoid a lot of conflicts and contention)

In short:
* Please rebase/cherry-pick so this is directly off of main
* Let's merge it
* Start a new PR adding native-staking-proxy (without tests)
* When vault lands (with proof of working Sylvia multi-test reply helpers), make new PRs adding tests

File: contracts/native-staking/src/contract.rs (hunk 10 above)
Comment (maurolacy) [reply contracts/native-staking/src/contract.rs:175]: The reply handler needs it (so as to properly populate the maps). But good point, we don't need to set here, but in the staking proxy instantiate return message. And it can get it from the `owner` field, above. Simpler is better.

**Comment by maurolacy** [comment]
> It is unfortunate, that this is based on vault. It would be good to get smaller independent PRs merged often to main.

I did just to take a look / copy code from it.
 
> 
> * Please rebase/cherry-pick so this is directly off of main

Just created a fresh branch from `main`, and cherry picked the relevant commit range.

> * Let's merge it
> * Start a new PR adding native-staking-proxy (without tests)
> * When vault lands (with proof of working Sylvia multi-test reply helpers), make new PRs adding tests

👍🏼

**Comment by maurolacy** [comment]
Was taking a look at the CI errors, and can't make sense of them. Do we merge anyway? Will take a closer look tomorrow, in any case.

**Comment by maurolacy** [comment]
> Was taking a look at the CI errors, and can't make sense of them. Do we merge anyway? Will take a closer look tomorrow, in any case.

Solved it by `cargo update`ing sylvia.

**Comment by ethanfrey** [comment]
Great, thanks for the cleanup.
I will rebase on top.
